lib/msf/core/auxiliary/masscan.rb

```diff
@@ -0,0 +1,277 @@
+# -*- coding: binary -*-
+require 'rex/parser/masscan_nokogiri'
+require 'rex/parser/masscan_xml'
+require 'open3'
+
+module Msf
+
+###
+#
+# This module provides methods for interacting with nmap.
+# Modules that include this should define their own nmap_build_args()
+# function, and usually should have some method for dealing with
+# the data yielded from nmap_hosts(). See auxiliary/scanner/oracle/oracle_login
+# for an example implementation.
+#
+###
+
+module Auxiliary::Masscan
+
+attr_accessor :massscan_args, :masscan_bin, :masscan_log
+attr_reader :masscan_pid, :masscan_ver
+
+def initialize(info = {})
+  super
+
+  register_options([
+    OptAddressRange.new('RHOSTS', [ true, "The target address range or CIDR identifier"]),
+    OptBool.new('NMAP_VERBOSE', [ false, 'Display nmap output', true]),
+    OptString.new('RPORTS', [ false, 'Ports to target']), # RPORT supersedes RPORTS
+  ], Auxiliary::Nmap)
+
+  deregister_options("RPORT")
+  @masscan_args = []
+  @masscan_bin = masscan_binary_path
+end
+
+def rports
+  datastore['RPORTS']
+end
+
+def rport
+  datastore['RPORT']
+end
+
+def set_masscan_cmd
+  self.masscan_bin || (raise RuntimeError, "Cannot locate massscan binary")
+  masscan_set_log
+  masscan_add_ports
+  masscan_cmd = [self.nmap_bin]
+  self.masscan_args.unshift("-oX #{self.masscan_log[1]}")
+  masscan_cmd << self.masscan_args.join(" ")
+  masscan_cmd << datastore['RHOSTS']
+  masscan_cmd.join(" ")
+end
+
+def get_masscan_ver
+  self.masscan_bin || (raise RuntimeError, "Cannot locate nmap binary")
+  res = ""
+  masscan_cmd = [self.masscan_bin]
+  masscan_cmd << "--version"
+  res << %x{#{masscan_cmd.join(" ")}} rescue nil
+  res.gsub(/[\x0d\x0a]/n,"")
```

Comment:
This will fail if the execution of masscan fails, because `res` will then be `nil` and you can't call any `gsub` on `nil`